cfme/tests/distributed/test_appliance.py

```diff
@@ -0,0 +1,198 @@
+            view.save.click()
+            view.flash.assert_no_error()
+
+            wait_for(lambda: secondary_appliance.httpd.running, delay=10)
+
+
+@pytest.mark.ignore_stream("upstream")
+def test_appliance_reporting_role(distributed_appliances):
+    """Test that a report queued from an appliance with the User Interface role but not the
+    Reporting role gets successfully run by a worker appliance that does have the Reporting
+    role.
+
+    Bugzilla:
+        1629945
+
+    Polarion:
+        assignee: tpapaioa
+        casecomponent: Appliance
+        caseimportance: medium
+        initialEstimate: 1/4h
+    """
+    primary_appliance, secondary_appliance = distributed_appliances
+
+    # Disable the Reporting role on the primary appliance.
+    primary_appliance.server.settings.disable_server_roles('reporting')
+
+    # Wait for the role to be disabled in the database.
+    wait_for(lambda: not primary_appliance.server.settings.server_roles_db['reporting'])
+
+    # Queue the report and wait for it to complete.
+    primary_appliance.collections.reports.instantiate(
+        type="Operations",
+        subtype="EVM",
+        menu_name="EVM Server UserID Usage Report"
+    ).queue(wait_for_finish=True)
+
+
+@pytest.mark.ignore_stream('upstream')
+def test_server_role_failover(distributed_appliances):
+    """Test that server roles failover successfully to a secondary appliance if evmserverd stops
+    on the primary appliance.
+
+    Polarion:
+        assignee: tpapaioa
+        casecomponent: Appliance
+        caseimportance: medium
+        initialEstimate: 1/12h
+    """
+    primary_appliance, secondary_appliance = distributed_appliances
+
+    all_server_roles = cfme_data.get('server_roles', {'all': []})['all']
+    if not all_server_roles:
+        pytest.skip('Empty server_roles dictionary in cfme_data, skipping test')
+
+    # Remove roles in cfme_data that are not in 5.11 or later.
+    remove_roles = ['websocket']
+    server_roles = [r for r in all_server_roles if r not in remove_roles]
+    fill_values = {k: True for k in server_roles}
+
+    # Enable all roles on both appliances.
+    for appliance in distributed_appliances:
+        appliance.browser_steal = True
+        with appliance:
+            view = navigate_to(appliance.server, 'Server')
+            view.server_roles.fill(fill_values)
+            view.save.click()
+            view.flash.assert_no_error()
+
+    # Stop evmserverd on secondary appliance.
+    secondary_appliance.evmserverd.stop()
+
+    # Verify that all roles are active on primary appliance.
+    wait_for(lambda: primary_appliance.server_roles == fill_values)
+
+    # Stop evmserverd on primary appliance and restart it on secondary appliance.
+    secondary_appliance.evmserverd.start()
+    primary_appliance.evmserverd.stop()
+
+    # Verify that all roles are now active on secondary appliance.
+    wait_for(lambda: secondary_appliance.server_roles == fill_values)
```

Comment:
I am surprised you didn't hit an HTTP 503 exception here. I don't believe the API will be always up at this point.

I think you need to use https://github.com/ManageIQ/integration_tests/blob/master/cfme/utils/appliance/__init__.py#L1492 or wait_for_miq_ready

Reply:
Hmm, I think I understand. The `wait_for` is probably eating the exception.

I think you should add the `wait_for_api_available` anyway.

Reply:
@tpapaioa @jaryn PR adding the new appliance wait method has been merged, please use it here to ensure the evm service and all associated web services are available before interacting with the `secondary_appliance`